src/lsm/scan_groove.zig

```diff
@@ -0,0 +1,892 @@
+                    .decls = &.{},
+                },
+            };
+
+            // Union fields for each index tree:
+            for (std.meta.fields(Groove.IndexTrees)) |field| {
+                const IndexTree = field.type;
+                const ScanTree = ScanTreeType(*Context, IndexTree, Storage);
+                type_info.Union.fields = type_info.Union.fields ++
+                    [_]std.builtin.Type.UnionField{.{
+                    .name = field.name,
+                    .type = ScanTree,
+                    .alignment = @alignOf(ScanTree),
+                }};
+            }
+
+            // Add fields for set operations that share the same interface:
+            const ScanMergeUnion = ScanMergeUnionType(Scan);
+            type_info.Union.fields = type_info.Union.fields ++
+                [_]std.builtin.Type.UnionField{.{
+                .name = "merge_union",
+                .type = ScanMergeUnion,
+                .alignment = @alignOf(ScanMergeUnion),
+            }};
+            const ScanMergeIntersection = ScanMergeIntersectionType(Scan);
+            type_info.Union.fields = type_info.Union.fields ++
+                [_]std.builtin.Type.UnionField{.{
+                .name = "merge_intersection",
+                .type = ScanMergeIntersection,
+                .alignment = @alignOf(ScanMergeIntersection),
+            }};
+            const ScanMergeDifference = ScanMergeDifferenceType(Scan);
+            type_info.Union.fields = type_info.Union.fields ++
+                [_]std.builtin.Type.UnionField{.{
+                .name = "merge_difference",
+                .type = ScanMergeDifference,
+                .alignment = @alignOf(ScanMergeDifference),
+            }};
+
+            // We need a tagged union for dynamic dispatching.
+            type_info.Union.tag_type = blk: {
+                const union_fields = type_info.Union.fields;
+                var tag_fields: [union_fields.len]std.builtin.Type.EnumField =
+                    undefined;
+                for (union_fields, 0..) |union_field, i| {
+                    tag_fields[i] = .{
+                        .name = union_field.name,
+                        .value = i,
+                    };
+                }
+
+                break :blk @Type(.{ .Enum = .{
+                    .tag_type = std.math.IntFittingRange(0, tag_fields.len - 1),
+                    .fields = &tag_fields,
+                    .decls = &.{},
+                    .is_exhaustive = true,
+                } });
+            };
+
+            return @Type(type_info);
+        }
+    };
+}
+
+/// Union ∪ operation over an array of non-specialized `Scan` instances.
+fn ScanMergeUnionType(
+    comptime Scan: type,
+) type {
+    return struct {
+        const Self = @This();
+        pub const Callback = *const fn (context: *Scan.Context, self: *Self) void;
+
+        const BoundedArray = stdx.BoundedArray(struct {
+            scan: *Scan,
+            popped: ?u64,
```

Comment:
The `KWayMerge` interface returns `error { Drained, Empty }!Value` on `peek()` and `pop()` never fails, but it requires a prior call to `peek()`, otherwise it panics.
On the other hand, the `Scan` interface is intended to be used with the `while(scan.next()) |item| { }` pattern.

I don't think a common interface for both can be useful, but I do think we can improve the way the `Scan` is used as a KWayMerge source.